cluster-autoscaler/cloudprovider/azure/azure_agent_pool.go

```diff
@@ -117,9 +131,35 @@ func (as *AgentPool) MaxSize() int {
 	return as.maxSize
 }
 
+func (as *AgentPool) getAllVirtualMachinesWithCache() ([]compute.VirtualMachine, error) {
+	virtualMachinesStatusCache.mutex.Lock()
+	defer virtualMachinesStatusCache.mutex.Unlock()
+
+	if virtualMachinesStatusCache.lastRefresh.Add(vmInstancesRefreshPeriod).After(time.Now()) {
+		return virtualMachinesStatusCache.virtualMachines, nil
+	}
+
+	vms, err := as.GetVirtualMachines()
```

Comment:
should check err before the cache update. or else, the cache may be cleaned on errors